### ISSUE TITLE
Stuff in MSSQL

### ISSUE THREAD

**Issue by leibale** [sequelize/sequelize#5676]
use `STUFF` function and `FOR` Clause to enhance `hasMany` relation.
[STUFF in MSDN](https://msdn.microsoft.com/en-us/library/ms188043.aspx), [FOR in MSDN](https://msdn.microsoft.com/en-us/library/ms173812.aspx)

example:
```sql
SELECT Id, STUFF((SELECT Id FROM Table2 WHERE Table2.Table1Id = Table1.Id FOR JSON AUTO), 1, 0, '') AS items FROM Table2
```

**Comment by janmeier** [comment]
I'm not sure what benefit this would have over a join? 

Generally we try not to support too many features which only exist in a single dialect

**Comment by leibale** [comment]
Table1

| Id | Name |
| --- | --- |
| 1 | Test |

Table2

| Id | Table1Id |
| --- | --- |
| 1 | 1 |
| 2 | 1 |

if you are using a join the result from the db will be

| Table1.Id | Table1.Name | Table2.Id | Table2.Table1Id |
| --- | --- | --- | --- |
| 1 | Test | 1 | 1 |
| 1 | Test | 2 | 1 |

and if you are using `STUFF` and `FOR` the result will be

| Table1.Id | Table1.Name | items |
| --- | --- | --- |
| 1 | Test | [{Id: 1, Table1Id: 1}, {Id: 2, Table1Id: 1}] |

`JOIN` duplicates the left side, which in real world scenario can translate to a lot of data.

**Comment by leibale** [comment]
the alternativ in `mysql` is `GROUP_CONCAT` function.
http://dev.mysql.com/doc/refman/5.5/en/group-by-functions.html#function_group-concat

**Comment by mickhansen** [comment]
`array_to_json` and `array_agg` might be usefull for postgres.

**Comment by mickhansen** [comment]
We need to consider how this would look with nested includes.
We would also need to do extensive benchmarking of different cases to document a real (atleast 20%) performance gain to consider a rewrite.

**Comment by mickhansen** [comment]
This would also remove the possibility of filtering main rows based on a join.

**Comment by gkorland** [comment]
+1

**Comment by leibale** [comment]
@mickhansen why do you believe we won't be able to filter main rows?

**Comment by mickhansen** [comment]
@leibale I thought about it again and i suppose you still can, in some cases. With stuff/group_concat won't the column be a string? You won't be able to match on the string.

**Comment by leibale** [comment]
@mickhansen the result will be string, but in the db you still can use it like a field so you still can filter by whatever you want.

**Comment by mickhansen** [comment]
@leibale How would you target a column across multiple includes inside that string though? I'm genuinely asking, would love to see that expressed in SQL.

**Comment by leibale** [comment]
@mickhansen 
```sql
SELECT Id id, Name name,
  STUFF((SELECT Id id, Table1Id table1Id FROM Table2 WHERE Table1Id = Table1.Id FOR JSON AUTO), 1, 0, '')
FROM Table1
```
and if you want to filter `Table1` by `Table2` you just need to do something like that:
```sql
SELECT DISTINCT Table1.Id id, Table1.Name name,
STUFF((SELECT Id id, Table1Id table1Id FROM Table2 WHERE Id = Table2.Id FOR JSON AUTO), 1, 0, '')
FROM Table1
LEFT JOIN Table2 ON Table2.Table1Id = Table1.Id
WHERE Table2.Id = 2
```

**Comment by mickhansen** [comment]
@leibale In that case you have both a subquery and a join for the same table.

**Comment by leibale** [comment]
@mickhansen yes, you right. in this case i think using `JOIN` is fine. but the most of the cases I need to use `STUFF` . maybe the best is to implement `STUFF` also and use it only if the user not filter the main rows by the join.

**Comment by mickhansen** [comment]
That would be possible, we'd need to maintain two seperate code paths and that requires more testing.
It's unlikely that the core team will look at this anytime soon, but it could be a nice optimization in the future if we can implement something cross dialect that improves performance.

**Comment by leibale** [comment]
@mickhansen in the most cases user not filter the main rows based on the join, so i think you should first implement only `STUFF` and then add a "nice optimization" by adding the `JOIN` again.

**Comment by mickhansen** [comment]
@leibale A lot of cases they do, seeing as we often have issues about it :) Eitherway it does not change the harder maintenance/harder testing issue.

**Comment by leibale** [comment]
@mickhansen if i will create a PR you will help me to merge?

**Comment by mickhansen** [comment]
I'll definitely take a look @leibale, can't promise that we'll merge since it's additional maintenance for us.
A PR would also need to provide an implementation for all dialects (psql, mysql, sqlite, mssql).

**Comment by leibale** [comment]
@mickhansen so.. i will wait for you..

**Comment by leibale** [comment]
@mickhansen ...

**Comment by mickhansen** [comment]
@leibale I'm not actively involved with Sequelize anymore.

**Comment by leibale** [comment]
@mickhansen can you please tag someone that "actively involved with Sequelize"?

**Comment by leibale** [comment]
It is still a an issue

**Comment by leibale** [comment]
It is still an issue

**Comment by gkorland** [comment]
The issue is still open
